pkg/kubernetes/deployment/appsv1/v1alpha1/kubernetes.go

```diff
@@ -0,0 +1,136 @@
+package v1alpha1
+
+import (
+	client "github.com/openebs/maya/pkg/kubernetes/client/v1alpha1"
+	api_apps_v1 "k8s.io/api/apps/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// getClientsetFn is a typed function that abstracts fetching of internal clientset
+type getClientsetFn func() (clientset *kubernetes.Clientset, err error)
+
+// getFn is a typed function that abstracts get of deployment instances
+type getFn func(cli *kubernetes.Clientset, name, namespace string) (*api_apps_v1.Deployment, error)
+
+// rolloutStatusFn is a typed function that abstracts rollout status of deployment instances
+type rolloutStatusFn func(cli *kubernetes.Clientset, name, namespace string) ([]byte, error)
+
+// kubeclient enables kubernetes API operations on deployment instance
+type kubeclient struct {
+	// clientset refers to kubernetes clientset. It is responsible to
+	// make kubernetes API calls for crud op
+	clientset *kubernetes.Clientset
+	namespace string
+
+	// functions useful during mocking
+	getClientset  getClientsetFn
+	get           getFn
+	rolloutStatus rolloutStatusFn
+}
+
+// rolloutOutput struct contaons message and boolean value to show rolloutstatus
+type rolloutOutput struct {
+	IsRolledout bool   `json:"isRolledout"`
+	Message     string `json:"message"`
+}
+
+// kubeclientBuildOption defines the abstraction to build a kubeclient instance
+type kubeclientBuildOption func(*kubeclient)
+
+// withDefaults sets the default options of kubeclient instance
+func (k *kubeclient) withDefaults() {
+	if k.getClientset == nil {
+		k.getClientset = func() (clients *kubernetes.Clientset, err error) {
+			config, err := client.GetConfig(client.New())
+			if err != nil {
+				return nil, err
+			}
+			return kubernetes.NewForConfig(config)
+		}
+	}
+
+	if k.get == nil {
+		k.get = func(cli *kubernetes.Clientset, name,
+			namespace string) (d *api_apps_v1.Deployment, err error) {
+			d = &api_apps_v1.Deployment{}
+			err = cli.ExtensionsV1beta1().
+				RESTClient().
+				Get().
+				Namespace(namespace).
+				Name(name).
+				Resource("deployments").
+				Do().
+				Into(d)
+			return
+		}
+	}
+
+	if k.rolloutStatus == nil {
+		k.rolloutStatus = func(cli *kubernetes.Clientset, name,
+			namespace string) (op []byte, err error) {
+			d, err := k.get(cli, name, namespace)
```

Comment:
Let us move k.get out of k.rolloutStatus
Let rolloutStatusFn be concerned with rollout only. let it accept a deployment object if reqd.

Reply:
Alternatively, IMO we should avoid use of `*k` pointer inside these default implementations

Reply:
done please check it once.